### PR TITLE
Stax port

### PR DESCRIPTION
**Pull request by agrojean-ledger**
### Port Tron app to Stax
* Refactoring of UI code : move code from main.c to UI specific files.
* Add Stax UI (main menu with settings + tx review menu).
* Remove obsolete Blue specific code.

### CODE REVIEW

File: src/ui/ui_idle_menu_nbgl.c [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    SWITCH_ALLOW_HASH_TX_TOKEN
+};
+
+#define NB_INFO_FIELDS 2
+static const char* const infoTypes[] = {"Version", "Tron App"};
+static const char* const infoContents[] = {APPVERSION, "(c) 2022 Ledger"};
+
+#define NB_SETTINGS_SWITCHES 3
+#define SWITCH_IDX(token) (token - SWITCH_ALLOW_TX_DATA_TOKEN)
+static uint8_t settings[NB_SETTINGS_SWITCHES] = {S_DATA_ALLOWED,S_CUSTOM_CONTRACT,S_SIGN_BY_HASH};
```
Comment (xchapron-ledger) [inline src/ui/ui_idle_menu_nbgl.c:40]:
```suggestion
static uint8_t settings[NB_SETTINGS_SWITCHES] = {S_DATA_ALLOWED, S_CUSTOM_CONTRACT, S_SIGN_BY_HASH};
```

File: src/ui/ui_idle_menu_nbgl.c [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+static void displaySettingsMenu(void) {
+    nbgl_useCaseSettings("Tron settings",0,2,true,ui_idle,settingsNavCallback,settingsControlsCallback);
+}
+
+void ui_idle(void) {
+    switches[SWITCH_IDX(SWITCH_ALLOW_TX_DATA_TOKEN)].initState = (HAS_SETTING(S_DATA_ALLOWED))?ON_STATE:OFF_STATE;
+    switches[SWITCH_IDX(SWITCH_ALLOW_CSTM_CONTRACTS_TOKEN)].initState = (HAS_SETTING(S_CUSTOM_CONTRACT))?ON_STATE:OFF_STATE;
+    switches[SWITCH_IDX(SWITCH_ALLOW_HASH_TX_TOKEN)].initState = (HAS_SETTING(S_SIGN_BY_HASH))?ON_STATE:OFF_STATE;
```
Comment (xchapron-ledger) [inline src/ui/ui_idle_menu_nbgl.c:107]: Would be better in `settingsNavCallback()`? Or at least in `displaySettingsMenu()`.
Also can you add spaces.

File: src/ui/ui_idle_menu_nbgl.c [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            displaySettingsMenu();
+            break;    
+        default:
+            PRINTF("Should not happen !");
+            break;
+    }
+}
+
+static void displaySettingsMenu(void) {
+    nbgl_useCaseSettings("Tron settings",0,2,true,ui_idle,settingsNavCallback,settingsControlsCallback);
```
Comment (xchapron-ledger) [inline src/ui/ui_idle_menu_nbgl.c:101]: Please add spaces.

File: src/ui/ui_review_menu_bagl.c [hunk 4]
```diff
@@ -0,0 +1,859 @@
...
+      &C_icon_certificate,
+      "Review",
+      "Transaction",
+    });
+UX_STEP_NOCB(
+    ux_approval_tx_2_step,
+    bnnn_paging,
+    {
+      .title = "Amount",
+      .text = (char *) G_io_apdu_buffer
```
Comment (xchapron-ledger) [inline src/ui/ui_review_menu_bagl.c:147]: Doesn't seem great, but I guess not from you?

File: src/ui/ui_review_menu_bagl.c [hunk 5]
```diff
@@ -0,0 +1,859 @@
...
+    {
+      .title = "Token 2",
+      .text = toAddress,
+    });
+UX_STEP_NOCB(
+    ux_approval_exchange_create_5_step,
+    bnnn_paging,
+    {
+      .title = "Amount 2",
+      .text = (char *) G_io_apdu_buffer + 100,
```
Comment (xchapron-ledger) [inline src/ui/ui_review_menu_bagl.c:221]: :cry:

File: src/ui/ui_idle_menu_nbgl.c [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+ *  limitations under the License.
+ ********************************************************************************/
+#ifdef HAVE_NBGL
+
+#include "os.h"
+#include "os_io_seproxyhal.h"
+#include "ui_idle_menu.h"
+#include "glyphs.h"
+#include "ux.h"
+#include "nbgl_page.h"
```
Comment (xchapron-ledger) [inline src/ui/ui_idle_menu_nbgl.c:24]: not needed?

File: src/ui/ui_review_menu_nbgl.c [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+ *  limitations under the License.
+ ********************************************************************************/
+#ifdef HAVE_NBGL
+#include <stdbool.h>
+#include <sys/types.h>
+#include <string.h>
+#include <stdint.h>
+
+#include "ux.h"
+#include "nbgl_page.h"
```
Comment (xchapron-ledger) [inline src/ui/ui_review_menu_nbgl.c:24]: needed?

File: src/ui/ui_review_menu_nbgl.c [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+#include "nbgl_use_case.h"
+#include "os_io_seproxyhal.h"
```
Comment (xchapron-ledger) [inline src/ui/ui_review_menu_nbgl.c:26]: needed?

File: src/ui/ui_review_menu_nbgl.c [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+}
+
+static void prepareTxInfos(ui_approval_state_t state, bool data_warning) {
+    memset(&txInfos, 0, sizeof(txInfos));
+    memset(&infoLongPress, 0, sizeof(infoLongPress));
+
+    txInfos.warnings[DATA_WARNING] = data_warning;
+    txInfos.flowTitle = "Review Transaction";
+    txInfos.confirmCb = (nbgl_callback_t) io_seproxyhal_touch_tx_ok;
+    txInfos.rejectCb = (nbgl_callback_t) io_seproxyhal_touch_cancel;
```
Comment (xchapron-ledger) [inline src/ui/ui_review_menu_nbgl.c:154]: If common, maybe they don't need to be in `txInfos`

File: src/ui/ui_review_menu_nbgl.c [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+            txInfos.fields[2].value = fullContract;
+            txInfos.fields[3].item = "Amount";
+            txInfos.fields[3].value = (char *) G_io_apdu_buffer;
+            txInfos.fields[4].item = "From Address";
+            txInfos.fields[4].value = fromAddress;
+            pairList.nbPairs = 5;
+            break;
+        case APPROVAL_WITNESSVOTE_TRANSACTION:
+            uint8_t i;
+            for (i = 0; i < votes_count; i++) {
```
Comment (xchapron-ledger) [inline src/ui/ui_review_menu_nbgl.c:235]: add a check on size of `txInfos.fields` array?

File: src/ui/ui_idle_menu_nbgl.c (hunk 1 above)
Comment (agrojean-ledger) [reply src/ui/ui_idle_menu_nbgl.c:40]: @xchapron-ledger there is a dedicated format PR : https://github.com/LedgerHQ/app-tron/pull/42/

File: src/ui/ui_idle_menu_nbgl.c (hunk 2 above)
Comment (agrojean-ledger) [reply src/ui/ui_idle_menu_nbgl.c:107]: Regarding spaces, https://github.com/LedgerHQ/app-tron/pull/42/ @xchapron-ledger

File: src/ui/ui_review_menu_bagl.c (hunk 4 above)
Comment (agrojean-ledger) [reply src/ui/ui_review_menu_bagl.c:147]: @xchapron-ledger not from me at all. Legacy. Don't want to refacto the whole architecture...

File: src/ui/ui_review_menu_bagl.c (hunk 5 above)
Comment (agrojean-ledger) [reply src/ui/ui_review_menu_bagl.c:221]: Not from me either @xchapron-ledger ...

File: src/ui/ui_idle_menu_nbgl.c (hunk 6 above)
Comment (agrojean-ledger) [reply src/ui/ui_idle_menu_nbgl.c:24]: I'll remove it.

File: Makefile [hunk 11]
```diff
@@ -92,17 +94,22 @@ DEFINES   += BLE_SEGMENT_SIZE=32 #max MTU, min 20
...
+DEFINES   += HAVE_BLE BLE_COMMAND_TIMEOUT_MS=2000 HAVE_BLE_APDU
+else ifeq ($(TARGET_NAME),TARGET_STAX)
+DEFINES   += HAVE_BLE BLE_COMMAND_TIMEOUT_MS=2000 HAVE_BLE_APDU
 endif
 
 ifeq ($(TARGET_NAME),TARGET_NANOS)
 DEFINES   += IO_SEPROXYHAL_BUFFER_SIZE_B=128
+else ifeq ($(TARGET_NAME),TARGET_STAX)
+DEFINES       += IO_SEPROXYHAL_BUFFER_SIZE_B=300
+DEFINES       += HAVE_NBGL
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply Makefile:108]: Ok.

File: src/ui/ui_review_menu_nbgl.c (hunk 9 above)
Comment (agrojean-ledger) [reply src/ui/ui_review_menu_nbgl.c:154]: @xchapron-ledger actually the callbacks can be different depending on the tx type : https://github.com/LedgerHQ/app-tron/blob/57f7a8ea36cdf6fb719ecacc5ba2d866dab02970/src/ui/ui_review_menu_nbgl.c#L313

So I use the structure to keep the correct one for later use.

File: src/ui/ui_review_menu_nbgl.c (hunk 7 above)
Comment (agrojean-ledger) [reply src/ui/ui_review_menu_nbgl.c:24]: No, I'll remove it.

File: src/ui/ui_review_menu_nbgl.c (hunk 8 above)
Comment (agrojean-ledger) [reply src/ui/ui_review_menu_nbgl.c:26]: Nope, I'll remove it.

File: src/ui/ui_globals.h [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+extern char fullHash[HASH_SIZE*2+1];
+extern int8_t votes_count;
+extern transactionContext_t transactionContext;
+extern publicKeyContext_t publicKeyContext;
+
+unsigned int io_seproxyhal_touch_tx_ok(const void *e);
+unsigned int io_seproxyhal_touch_cancel(const void *e);
+unsigned int io_seproxyhal_touch_address_ok(const void *e);
+unsigned int io_seproxyhal_touch_signMessage_ok(const void *e);
+unsigned int io_seproxyhal_touch_ecdh_ok(const void *e);
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ui/ui_globals.h]: Done.

File: Makefile (hunk 11 above)
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply Makefile]: Done.

File: .gitignore [hunk 13]
```diff
@@ -38,6 +38,7 @@ glyphs.h
 *.x86_64
 *.hex
 bin/**
+build/
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply .gitignore]: I cleaned the gitignore in the tests PR as you requested : https://github.com/LedgerHQ/app-tron/blob/e0e30b8fa0c2fdd40ca3270cf6980f517b571557/.gitignore

File: src/main.c [hunk 14]
```diff
@@ -70,2621 +70,37 @@ uint32_t set_result_get_publicKey(void);
 #define OFFSET_LC 4
 #define OFFSET_CDATA 5
 
+bolos_ux_params_t G_ux_params;
+ux_state_t G_ux;
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/main.c:74]: Done.

File: src/main.c [hunk 15]
```diff
@@ -3712,7 +1020,9 @@ unsigned char io_event(unsigned char channel) {
         break;
 
     case SEPROXYHAL_TAG_DISPLAY_PROCESSED_EVENT:
+#ifdef HAVE_BAGL
         UX_DISPLAYED_EVENT({});
+#endif // HAVE_BAGL
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/main.c:1025]: Don't know what it's used for but it's in the boilerplate so I guess you're right. I'll add it.

File: src/ui/ui_idle_menu_bagl.c [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+UX_DEF(ux_idle_flow,
+  &ux_idle_flow_1_step,
+  &ux_idle_flow_2_step,
+  &ux_idle_flow_3_step,
+  &ux_idle_flow_4_step
+);
+#endif // HAVE_UX_FLOW
+
+void ui_idle(void) {
+#if defined(HAVE_UX_FLOW)
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ui/ui_idle_menu_bagl.c:207]: Good point.

File: src/ui/ui_idle_menu_bagl.c [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+ *  limitations under the License.
+ ********************************************************************************/
+#ifdef HAVE_BAGL
+#include "os.h"
+#include "os_io_seproxyhal.h"
+#include "ui_idle_menu.h"
+#include "ux.h"
+#include "settings.h"
+
+#if defined(HAVE_UX_FLOW)
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/ui/ui_idle_menu_bagl.c:24]: Indeed.

File: src/main.c [hunk 18]
```diff
@@ -3697,7 +1003,9 @@ unsigned char io_event(unsigned char channel) {
         break;
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/main.c:1002]: Done.

File: src/ui/ui_review_menu_nbgl.c (hunk 10 above)
Comment (agrojean-ledger) [reply src/ui/ui_review_menu_nbgl.c:237]: Done.

File: src/ui/ui_review_menu_nbgl.c [hunk 19]
```diff
@@ -0,0 +1,335 @@
...
+        nbgl_useCaseStatus("TRANSACTION\nCONFIRMED", true, ui_idle);
+        txInfos.confirmCb();
+    } else {
+        rejectChoice();
+    }
+}
+
+static void rejectConfirmation(void) {
+    nbgl_useCaseStatus("Transaction rejected", false, ui_idle);
+    txInfos.rejectCb();
```
Comment (xchapron-ledger) [inline src/ui/ui_review_menu_nbgl.c:136]: Sometimes you display the status then call the callback, and sometimes you do it the other way (in `customContractWarningChoice` for example).
Can you harmonize it?

File: src/ui/ui_review_menu_nbgl.c [hunk 20]
```diff
@@ -0,0 +1,335 @@
...
+                THROW(E_INCORRECT_DATA);
+            }
+            for (i = 0; i < votes_count; i++) {
+                txInfos.fields[i].item = (char *) (G_io_apdu_buffer + voteSlot(i, VOTE_ADDRESS));
+                txInfos.fields[i].value = (char *) (G_io_apdu_buffer + voteSlot(i, VOTE_AMOUNT));
+            }
+            txInfos.fields[i].item = "Total Vote Count";
+            txInfos.fields[i].value = fullContract;
+            txInfos.fields[i + 1].item = "From Address";
+            txInfos.fields[i + 1].value = fromAddress;
```
Comment (xchapron-ledger) [inline src/ui/ui_review_menu_nbgl.c:245]:
```suggestion
            txInfos.fields[votes_count].item = "Total Vote Count";
            txInfos.fields[votes_count].value = fullContract;
            txInfos.fields[votes_count + 1].item = "From Address";
            txInfos.fields[votes_count + 1].value = fromAddress;
```
I think that it makes the added check on fields array lenght easier to verify?
But up to you.

Comment (agrojean-ledger) [reply src/ui/ui_review_menu_nbgl.c:245]: I don't understand

File: src/ui/ui_review_menu_nbgl.c (hunk 19 above)
Comment (agrojean-ledger) [reply src/ui/ui_review_menu_nbgl.c:136]: Ok

File: src/ui/ui_idle_menu_nbgl.c (hunk 2 above)
Comment (xchapron-ledger) [reply src/ui/ui_idle_menu_nbgl.c:105]: Up on this one :)

Comment (agrojean-ledger) [reply src/ui/ui_idle_menu_nbgl.c]: Done.

File: .github/workflows/ci.yml [hunk 21]
```diff
@@ -52,4 +52,4 @@ jobs:
         env:
           CTEST_OUTPUT_ON_FAILURE: 1
         run: |
-          pytest tests --tb=short -v --device ${{ matrix.device }}
+          pytest tests --tb=short -v -s --device ${{ matrix.device }}
```
Comment (xchapron-ledger) [inline .github/workflows/ci.yml:55]: a bit too verbose?

File: src/ui/ui_idle_menu_nbgl.c [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+        switches[0].token = SWITCH_ALLOW_TX_DATA_TOKEN;
+        switches[0].tuneId = TUNE_TAP_CASUAL;
+        switches[1].text = "Custom contracts";
+        switches[1].subText = "Allow unverified contracts";
+        switches[1].token = SWITCH_ALLOW_CSTM_CONTRACTS_TOKEN;
+        switches[1].tuneId = TUNE_TAP_CASUAL;
+        switches[2].text = "Sign by Hash";
+        switches[2].subText = "Allow hash-only transactions";
+        switches[2].token = SWITCH_ALLOW_HASH_TX_TOKEN;
+        switches[2].tuneId = TUNE_TAP_CASUAL;
```
Comment (xchapron-ledger) [inline src/ui/ui_idle_menu_nbgl.c:62]: what do you think of:
```suggestion
        switches[0].text = "Transactions data";
        switches[0].subText = "Allow extra data in transactions";
        switches[0].token = SWITCH_ALLOW_TX_DATA_TOKEN;
        switches[0].initState =  (HAS_SETTING(S_DATA_ALLOWED)) ? ON_STATE : OFF_STATE;
        switches[0].tuneId = TUNE_TAP_CASUAL;
        switches[1].text = "Custom contracts";
        switches[1].subText = "Allow unverified contracts";
        switches[1].token = SWITCH_ALLOW_CSTM_CONTRACTS_TOKEN;
        switches[1].initState =  (HAS_SETTING(S_CUSTOM_CONTRACT)) ? ON_STATE : OFF_STATE;
        switches[1].tuneId = TUNE_TAP_CASUAL;
        switches[2].text = "Sign by Hash";
        switches[2].subText = "Allow hash-only transactions";
        switches[2].token = SWITCH_ALLOW_HASH_TX_TOKEN;
        switches[2].initState =  (HAS_SETTING(S_SIGN_BY_HASH)) ? ON_STATE : OFF_STATE;
        switches[2].tuneId = TUNE_TAP_CASUAL;
```
To remove both:
- switch initial setting in  `displaySettingsMenu()`
- switch update in `settingsControlsCallback()`

File: src/ui/ui_idle_menu_nbgl.c [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+
+static void settingsControlsCallback(int token, uint8_t index) {
+    UNUSED(index);
+    switch (token) {
+        case SWITCH_ALLOW_TX_DATA_TOKEN:
+        case SWITCH_ALLOW_CSTM_CONTRACTS_TOKEN:
+        case SWITCH_ALLOW_HASH_TX_TOKEN:
+            switches[SWITCH_IDX(token)].initState = !(switches[SWITCH_IDX(token)].initState);
+            SETTING_TOGGLE(settings[SWITCH_IDX(token)]);
+            displaySettingsMenu();
```
Comment (xchapron-ledger) [inline src/ui/ui_idle_menu_nbgl.c:86]:
```suggestion
```
Not necessary

**Review by xchapron-ledger — APPROVED** [review]

File: src/ui/ui_idle_menu_nbgl.c (hunk 22 above)
Comment (agrojean-ledger) [reply src/ui/ui_idle_menu_nbgl.c:62]: Good point. Let's do that.

File: .github/workflows/ci.yml (hunk 21 above)
Comment (agrojean-ledger) [reply .github/workflows/ci.yml]: For troubleshooting. I removed it.

File: .github/workflows/ci.yml [hunk 24]
```diff
@@ -27,9 +26,10 @@ jobs:
       fail-fast: false
       matrix:
         include:
-          - device: nanos
-          - device: nanox
-          - device: nanosp
+          # - device: nanos
+          # - device: nanox
+          # - device: nanosp
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply .github/workflows/ci.yml]: Removed it.

**Pull request merged**